### PR TITLE
Batched matrix multiplication.

### PR DESCRIPTION
**Pull request by FOsterfeld**
split dimension is a batch dimension

## Due Diligence

- General:
    - [x]  **base branch** must be `main` for new features, latest release branch (e.g. `release/1.3.x`) for bug fixes
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [x] documentation updated where needed

## Description

Issue/s resolved: #890 

## Changes proposed:

-
-
-
-

#### Does this change modify the behaviour of other functions? If so, which?
yes / no

### CODE REVIEW

File: heat/core/linalg/basics.py [hunk 1]
```diff
@@ -487,9 +487,12 @@ def matmul(a: DNDarray, b: DNDarray, allow_resplit: bool = False) -> DNDarray:
     sanitation.sanitize_in(a)
     sanitation.sanitize_in(b)
 
-    if a.gshape[-1] != b.gshape[0]:
+    batch_dim = max(a.ndim, b.ndim) - 2
+    batched = batch_dim > 2
```
Comment (mrfh92) [inline heat/core/linalg/basics.py:491]: > 0 ?

**Comment by mrfh92** [comment]
Sieht gut aus soweit :+1:

Vorschläge für die weitere Arbeit: 
* Wenn du oben neben "codecov" (rotes X) auf Details und dann auf "View this Pull Request on Codecov" gehst, kannst du sehen, welche Zeilen deiner Änderungen nicht getestet werden bisher. Entsprechende Tests kannst du in `heat/core/linalg/tests/test_basics.py` unterbringen. 
Mit diesem Fall wäre dann gebatchte Matrizenmultiplikation für den Fall abgedeckt, dass die LA-Dimensionen nicht gesplitted sind. Dies entspricht quasi dem Hinzufügen von Batchdimensionen zum Fall "2x`split=None`" im ursprünglichen Code. 
* Als nächstes könnten wir den Fall angehen, dass zu den anderen 8 (?) bestehenden Fällen Batch-dimensionen hinzugefügt werden; nun wäre allerdings die Batchdimension nicht gesplitted und dafür evtl. die LA-Dimension. Mit etwas Glück lässt sich das recht einfach aus dem bestehenden Code erzeugen, da ja PyTorch bereits gut mit Batches arbeiten kann. 
* Den Fall, dass einmal eine LA-Dimension gesplitted ist und einmal eine Batchdimension gesplitted ist, würde ich ausschließen und eine enstprechende Fehlermeldung ausgeben.

File: heat/core/linalg/basics.py (hunk 1 above)
Comment (FOsterfeld) [reply heat/core/linalg/basics.py]: Thanks, you're right, it should be 0.

**Comment by FOsterfeld** [comment]
Regarding the possible memory issue of #360, I benchmarked the maximal memory usage of matmul for two n x n matrices in the possible split cases.

plot_matmul-heat_2

**Comment by FOsterfeld** [comment]
The same plot with log scaling in the y-axis: 

plot_matmul-heat_2

**Comment by mrfh92** [comment]
@FOsterfeld thx :+1:
@ClaudiaComito I think we can use this diagram as basis for further refactoring efforts of matmal in the future

**Comment by mrfh92** [comment]
@FOsterfeld For now, I would suggest to add a hint to the docs that in terms of memory consumption the combinations 01, 00, 11 (and 0None, None1) seem to perform quite good, whereas the split combinations 10, None0 and 1None should be avoided due to comparably high memory consumption.

**Review by mtar — CHANGES_REQUESTED** [review]
I realised that the additions are a copy and paste of the existing code for the most part. Can you incorporate the changes or reuse it instead?

**Comment by FOsterfeld** [comment]
@mrfh92 Here is the plot of the runtimes from the last benchmark:
plot_matmul-heat-runtime_2`

**Comment by FOsterfeld** [comment]
@mrfh92 Here are plots of the runtime and memory usage of the the bad split combinations and the case that we do resplit instead. Specifically, we resplit 1-0 to 0-0, 1-None to 1-1 and None-0 to 0-0.

plot_matmul-resplit-runtime_2

plot_matmul-resplit-memory_2

**Comment by mrfh92** [comment]
@FOsterfeld thx for the plots :+1:

File: heat/core/linalg/basics.py [hunk 2]
```diff
@@ -426,20 +426,24 @@ def matmul(a: DNDarray, b: DNDarray, allow_resplit: bool = False) -> DNDarray:
...
     allow_resplit : bool, optional
         Whether to distribute ``a`` in the case that both ``a.split is None`` and ``b.split is None``.
         Default is ``False``. If ``True``, if both are not split then ``a`` will be distributed in-place along axis 0.
 
     Notes
     -----
-    - If ``a`` is a split vector then the returned vector will be of shape (:math:`1xQ`) and will be split in the 1st dimension
-    - If ``b`` is a vector and either ``a`` or ``b`` is split, then the returned vector will be of shape (:math:`Lx1`) and will be split in the 0th dimension
+    - For batched inputs, batch dimensions (and possible splits along these axes) must coincide.
+    - If ``a`` is a split vector then the returned vector will be of shape (:math:`1xQ`) and will be split in the 1st dimension. If ``b`` is a vector and either ``a`` or ``b`` is split, then the returned vector will be of shape (:math:`Lx1`) and will be split in the 0th dimension. Analogous conventions apply to batched inputs.
```
[the remark this answers is not in the document]
Comment (FOsterfeld) [reply heat/core/linalg/basics.py:445]: @mtar How can I get this preview? On read the docs I only found the docs for the releases.

[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/linalg/basics.py]: The documentation describes in the part [Writing Heat Documentation](https://heat.readthedocs.io/en/stable/documentation_howto.html) how to build it locally.

**Comment by FOsterfeld** [comment]
> I realised that the additions are a copy and paste of the existing code for the most part. Can you incorporate the changes or reuse it instead?

Now there should be no more code duplications for the batched/unbatched cases.

**Comment by mrfh92** [comment]
@FOsterfeld Something seems to have changed that lets the tests in QR fail. Can you reproduce this error on the workstation?

**Review by JuanPedroGHM — APPROVED** [review]
:+1:

**Comment by mrfh92** [comment]
@mtar have your suggestions been addressed by the changes?

**Review by mtar — APPROVED** [review]

**Comment by JuanPedroGHM** [comment]
### Benchmarks results - Sponsored by [perun](https://github.com/Helmholtz-AI-Energy/perun)

| function                                |   mpi_ranks | device   | metric   |      value |   ref_value |         std |   % change | type            | alert   |   lower_quantile |   upper_quantile |
|-----------------------------------------|-------------|----------|----------|------------|-------------|-------------|------------|-----------------|---------|------------------|------------------|
| matmul_split_1                          |           4 | CPU      | RUNTIME  | 0.122906   |  0.109381   | 0.0313848   |   12.3656  | jump-detection  | True    |     nan          |     nan          |
| concatenate                             |           4 | CPU      | RUNTIME  | 0.167615   |  0.196439   | 0.0260386   |  -14.6728  | jump-detection  | True    |     nan          |     nan          |
| apply_inplace_normalizer                |           4 | CPU      | RUNTIME  | 0.00100724 |  0.00947982 | 8.56803e-05 |  -89.375   | jump-detection  | True    |     nan          |     nan          |
| matmul_split_0                          |           4 | GPU      | RUNTIME  | 0.0483279  |  0.0545347  | 0.0310459   |  -11.3814  | jump-detection  | True    |     nan          |     nan          |
| matmul_split_1                          |           4 | GPU      | RUNTIME  | 0.0367425  |  0.0272262  | 0.0292762   |   34.9526  | jump-detection  | True    |     nan          |     nan          |
| qr_split_1                              |           4 | CPU      | RUNTIME  | 0.179386   |  0.451767   | 0.00646491  |  -60.2924  | trend-deviation | True    |       0.179536   |       0.796189   |
| kmeans                                  |           4 | CPU      | RUNTIME  | 0.309642   |  0.909418   | 0.00239133  |  -65.9517  | trend-deviation | True    |       0.312577   |       1.79242    |
| reshape                                 |           4 | CPU      | RUNTIME  | 0.155737   |  0.365847   | 0.00110721  |  -57.431   | trend-deviation | True    |       0.156546   |       0.731662   |
| lanczos                                 |           4 | GPU      | RUNTIME  | 0.600791   |  0.586755   | 0.00316215  |    2.39205 | trend-deviation | True    |       0.579288   |       0.597313   |
| kmeans                                  |           4 | GPU      | RUNTIME  | 0.657029   |  0.632586   | 0.004188    |    3.86392 | trend-deviation | True    |       0.617689   |       0.64691    |
| kmedians                                |           4 | GPU      | RUNTIME  | 1.11586    |  1.01658    | 0.0125118   |    9.76642 | trend-deviation | True    |       0.994589   |       1.04602    |
| kmedoids                                |           4 | GPU      | RUNTIME  | 1.2525     |  1.14701    | 0.00847031  |    9.19655 | trend-deviation | True    |       1.12248    |       1.18111    |
| apply_inplace_robust_scaler_and_inverse |           4 | GPU      | RUNTIME  | 5.99905    |  5.60869    | 0.0446267   |    6.95995 | trend-deviation | True    |       5.47179    |       5.75939    |
| apply_inplace_normalizer                |           4 | GPU      | RUNTIME  | 0.00179942 |  0.00169872 | 4.94232e-05 |    5.92757 | trend-deviation | True    |       0.00166147 |       0.00175474 |

[Grafana Dashboard](http://57bc8d92-72f2-4869-accd-435ec06365cb.ka.bw-cloud-instance.org:3000/d/adjpqduq9r7k0a/heat-cb?orgId=1)
Last updated: 2024-09-06T06:12:05Z

**Pull request merged**